2.4 update of the cross-DAG dependencies guide (#430)
astronomer/airflow-guides: guides/cross-dag-dependencies.md

description: "How to implement dependencies between your Airflow DAGs."
date: 2021-06-07T00:00:00.000Z
slug: "cross-dag-dependencies"
tags: ["DAGs", "Subdags", "Dependencies", "Sensors", "Dataset"]
---

When designing Airflow DAGs, it is often best practice to put all related tasks in the same DAG. However, it's sometimes necessary to create dependencies between your DAGs. In this scenario, one node of a DAG is its own complete DAG, rather than just a single task. Throughout this guide, we'll use the following terms to describe DAG dependencies:

- Upstream DAG: A DAG that must reach a specified state before a downstream DAG can run
- Downstream DAG: A DAG that cannot run until an upstream DAG reaches a specified state

According to the Airflow documentation on [cross-DAG dependencies](https://airflow.apache.org/docs/apache-airflow/stable/howto/operator/external_task_sensor.html#cross-dag-dependencies), designing DAGs in this way can be useful when:

> - A DAG should only run after one or more datasets have been updated by tasks in other DAGs.

Reviewer: Move this point towards the end of the list because it's a little more complex of a use case?

> - A DAG should only run after one or more datasets have been updated by tasks in other DAGs.

Author: I... can't figure out what you mean I'm sorry 😅 The PR you linked are fixed to the Astro Azure docs page 🤔

- Airflow operators. See [Operators 101](https://www.astronomer.io/guides/what-is-an-operator/).
- Airflow sensors. See [Sensors 101](https://www.astronomer.io/guides/what-is-a-sensor/).

## Implementing Cross-DAG Dependencies

There are multiple ways to implement cross-DAG dependencies in Airflow, including:

- [Dataset driven scheduling](https://www.astronomer.io/guides/airflow-datasets/)
- The [`TriggerDagRunOperator`](https://registry.astronomer.io/providers/apache-airflow/modules/triggerdagrunoperator)
- The [`ExternalTaskSensor`](https://registry.astronomer.io/providers/apache-airflow/modules/externaltasksensor)
- The [Airflow API](https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html)

Which implementations you choose will depend on your DAG dependencies and Airflow setup. In this section, we detail how to use each method and ideal scenarios for each, as well as how to view dependencies in the Airflow UI.

> **Note:** It can be tempting to use SubDAGs to handle DAG dependencies, but we highly recommend against doing so as SubDAGs can create performance issues. Instead, use one of the other methods described below.

### Dataset driven scheduling

> In Airflow 2.4+, you can use datasets to create data driven dependencies between DAGs. This means that DAGs which access the same data can have explicit, visible relationships, and that DAGs can be scheduled based on updates to these datasets. Downstream DAGs can be made dependent on an arbitrary number of datasets to be updated.

Reviewer 2: ```suggestion
In Airflow 2.4+, you can use datasets to create data-driven dependencies between DAGs. This means that DAGs which access the same data can have explicit, visible relationships, and that DAGs can be scheduled based on updates to these datasets. Downstream DAGs can be made dependent on an arbitrary number of datasets to be updated.
```

> This method is useful when a downstream DAG should only run after one or more datasets have been updated by one or more upstream DAGs, especially if those updates can be very irregular. Additionally, you will have increased observability into the dependencies between your DAGs and datasets in the Airflow UI (see the 'DAG dependencies View' section).

Reviewer 2: ```suggestion
You should use this method if you have a downstream DAG that should only run after one or more datasets have been updated by one or more upstream DAGs, especially if those updates can be very irregular. Additionally, you will have increased observability into the dependencies between your DAGs and datasets in the Airflow UI (see the 'DAG dependencies View' section).
```

```Python
dataset1 = Dataset('s3://folder1/dataset_1.txt')

# producing task in the upstream DAG
EmptyOperator(
    task_id="producing_task",
    outlets=[dataset1]  # flagging to Airflow that dataset1 was updated
)
```

The downstream DAG is scheduled to run after `dataset1` has been updated by providing it to the `schedule` parameter.

```Python
dataset1 = Dataset('s3://folder1/dataset_1.txt')

# consuming DAG
with DAG(
    dag_id='consuming_dag_1',
    catchup=False,
    start_date=datetime.datetime(2022, 1, 1),
    schedule=[dataset1]
) as dag:
```

The downstream DAG's Next Run will show how many datasets it depends on and how many of those have been updated since the last DAG run. The screenshot below shows that the DAG `dataset_dependent_example_dag` is scheduled depending on two datasets, one of which has already been updated.

![DAG Dependencies View](https://assets2.astronomer.io/main/guides/cross-dag-dependencies/2_4_DatasetDependentDAG.png)

Check out the [Datasets and Data Driven Scheduling in Airflow](https://www.astronomer.io/guides/airflow-datasets/) guide to learn more and see an example implementation of this feature.

### TriggerDagRunOperator

> The `TriggerDagRunOperator` is an easy way to implement cross-DAG dependencies from within the upstream DAG. This operator allows you to have a task in one DAG that triggers another DAG in the same Airflow environment. Read more in-depth documentation about this operator on the [Astronomer Registry](https://registry.astronomer.io/providers/apache-airflow/modules/triggerdagrunoperator).

Author: o7

> The `TriggerDagRunOperator` is ideal for situations in which you want one upstream DAG to run up to a specific task, trigger a downstream DAG and then continue running the next task in the upstream DAG as soon as the downstream DAG has finished running. This behavior can be achieved by setting the parameters `wait_for_completion` of the `TriggerDagRunOperator` to `True` (`False` by default).

Reviewer 2: ```suggestion
The `TriggerDagRunOperator` is ideal for situations in which you want to trigger a downstream DAG in the middle of the upstream DAG, and then continue running the next task in the upstream DAG as soon as the downstream DAG has finished running. This behavior can be achieved by setting the parameters `wait_for_completion` of the `TriggerDagRunOperator` to `True` (`False` by default).
```

> A common use case is that the upstream DAG fetches new testing data for a machine learning pipeline, runs and tests a model and then publishes the model's prediction. In case of the model underperforming the `TriggerDagRunOperator` is used to kick off a DAG that retrains the model, while the upstream DAG waits. Once the model is retrained and tested by the downstream DAG the upstream DAG resumes and goes on to publish the new model's results.

Reviewer 2: ```suggestion
A common use case for this implementation is when an upstream DAG fetches new testing data for a machine learning pipeline, runs and tests a model, and publishes the model's prediction. In case of the model underperforming, the `TriggerDagRunOperator` is used to kick off a separate DAG that retrains the model, while the upstream DAG waits. Once the model is retrained and tested by the downstream DAG, the upstream DAG resumes and goes on to publish the new model's results.
```

    start_task >> trigger_dependent_dag >> end_task
```

In the following graph view, you can see that the `trigger_dependent_dag` task in the middle is the `TriggerDagRunOperator`, which runs the `dependent-dag`.

![Trigger DAG Graph](https://assets2.astronomer.io/main/guides/cross-dag-dependencies/trigger_dag_run_graph.png)

> Note that f your dependent DAG requires a config input or a specific execution date, these can be specified in the operator using the `conf` and `execution_date` params respectively.

Reviewer 2: ```suggestion
Note that if your dependent DAG requires a config input or a specific execution date, these can be specified in the operator using the `conf` and `execution_date` params respectively.
```

### ExternalTaskSensor

To create cross-DAG dependencies from within your downstream DAG, consider using one or more [`ExternalTaskSensor`s](https://registry.astronomer.io/providers/apache-airflow/modules/externaltasksensor). The downstream DAG will wait until a task is completed in the upstream DAG before moving on to the rest of the DAG.

> This method of creating cross-DAG dependencies is especially useful when you have a downstream DAG with different branches that depend on different tasks in one or more upstream DAGs to have succeeded. Instead of defining a whole DAG as being downstream of another DAG like with datasets, you can define specific tasks in a downstream DAG as having to wait for a specific task to finishe in an upstream DAG.

Reviewer 2: ```suggestion
This method of creating cross-DAG dependencies is especially useful when you have a downstream DAG with different branches that depend on different tasks in one or more upstream DAGs. Instead of defining an entire DAG as being downstream of another DAG like with datasets, you can define specific tasks in a downstream DAG as having to wait for a specific task to finish in an upstream DAG.
```

### DAG Dependencies View

In [Airflow 2.1](https://airflow.apache.org/docs/apache-airflow/stable/changelog.html#airflow-2-1-0-2021-05-21), a new cross-DAG dependencies view was added to the Airflow UI. This view shows all dependencies between DAGs in your Airflow environment as long as they are implemented using one of the following methods:

- Using a `TriggerDagRunOperator`
- Using an `ExternalTaskSensor`
- Using dependencies set on Datasets

Dependencies can be viewed in the UI by going to `Browse` → `DAG Dependencies` or by clicking on the `Graph` button from within the Datasets tab. The screenshot below shows the dependencies created by the `TriggerDagRunOperator` and `ExternalTaskSensor` example DAGs in the sections above.

![DAG Dependencies View](https://assets2.astronomer.io/main/guides/cross-dag-dependencies/dag_dependencies_view.png)

When DAGs are scheduled depending on Datasets, both the DAG containing the producing task, as well as the Dataset itself will be shown upstream of the consuming DAG.

![DAG Dependencies View Datasets](https://assets2.astronomer.io/main/guides/cross-dag-dependencies/2_4_CrossGuide_Dependencies.png)

> ## Cross-Deployment Dependencies

Author: As discussed the feature is only in the early stages of being made possible so I'm going to resolve the comment :)

![Http Connection](https://assets2.astronomer.io/main/guides/cross-dag-dependencies/http_connection.png)

## DAG Dependencies View

In [Airflow 2.1](https://airflow.apache.org/docs/apache-airflow/stable/changelog.html#airflow-2-1-0-2021-05-21), a new cross-DAG dependencies view was added to the Airflow UI. This view shows all dependencies between DAGs in your Airflow environment as long as they are implemented using one of the following methods:

- Using dataset driven scheduling
- Using a `TriggerDagRunOperator`
- Using an `ExternalTaskSensor`

Dependencies can be viewed in the UI by going to `Browse` → `DAG Dependencies` or by clicking on the `Graph` button from within the Datasets tab. The screenshot below shows the dependencies created by the `TriggerDagRunOperator` and `ExternalTaskSensor` example DAGs in the sections above.

![DAG Dependencies View](https://assets2.astronomer.io/main/guides/cross-dag-dependencies/dag_dependencies_view.png)

When DAGs are scheduled depending on datasets, both the DAG containing the producing task, as well as the dataset itself will be shown upstream of the consuming DAG.

> ![DAG Dependencies View Datasets](https://assets2.astronomer.io/main/guides/cross-dag-dependencies/2_4_CrossGuide_Dependencies.png)

Author: ```suggestion
![DAG Dependencies View Datasets](https://assets2.astronomer.io/main/guides/cross-dag-dependencies/2_4_CrossGuide_Dependencies.png)

In Airflow 2.4 an additional **Datasets** tab was added, which shows all dependencies between datasets and DAGs.
![DAG Dependencies View Datasets](https://assets2.astronomer.io/main/guides/cross-dag-dependencies/2_4_Datasets.png)
```

## DAG Dependencies View

In [Airflow 2.1](https://airflow.apache.org/docs/apache-airflow/stable/changelog.html#airflow-2-1-0-2021-05-21), a new cross-DAG dependencies view was added to the Airflow UI. This view shows all dependencies between DAGs in your Airflow environment as long as they are implemented using one of the following methods:

- Using dataset driven scheduling
- Using a `TriggerDagRunOperator`
- Using an `ExternalTaskSensor`

Dependencies can be viewed in the UI by going to `Browse` → `DAG Dependencies` or by clicking on the `Graph` button from within the Datasets tab. The screenshot below shows the dependencies created by the `TriggerDagRunOperator` and `ExternalTaskSensor` example DAGs in the sections above.

![DAG Dependencies View](https://assets2.astronomer.io/main/guides/cross-dag-dependencies/dag_dependencies_view.png)

When DAGs are scheduled depending on datasets, both the DAG containing the producing task, as well as the dataset itself will be shown upstream of the consuming DAG.

![DAG Dependencies View Datasets](https://assets2.astronomer.io/main/guides/cross-dag-dependencies/2_4_CrossGuide_Dependencies.png)

> In Airflow 2.4 an additional **Datasets** tab was added, which shows all dependencies between datasets and DAGs.

Author: ```suggestion
In Airflow 2.4 an additional **Datasets** tab was added, which shows all dependencies between datasets and DAGs.

```
